ouroboros-consensus-cardano/test/Test/Consensus/Cardano/Generators.hs

```diff
@@ -0,0 +1,340 @@
+
+arbitraryHardForkState
+  :: forall f sc a.
+     ( Arbitrary (f ByronBlock)
+     , Arbitrary (f (ShelleyBlock sc))
+     , Coercible a (HardForkState_ f f (CardanoEras sc))
+     )
+  => Proxy f
+  -> Gen a
+arbitraryHardForkState p = toTelescope' p <$> oneof
+    [ Left  <$> genCurrent (Proxy @ByronBlock)
+    , Right <$> ((,) <$> genPast    (Proxy @ByronBlock)
+                     <*> genCurrent (Proxy @(ShelleyBlock sc)))
+    ]
+  where
+    genCurrent
+      :: forall blk. Arbitrary (f blk)
+      => Proxy blk
+      -> Gen (Current f blk)
+    genCurrent _ = Current <$> arbitrary <*> (arbitrary @(f blk))
+
+    genPast
+      :: forall blk. Arbitrary (f blk)
+      => Proxy blk
+      -> Gen (Past f blk)
+    genPast _ = Past
+      <$> arbitrary
+      <*> arbitrary
+      <*> oneof
+            [ Snapshot <$> arbitrary <*> arbitrary @(f blk)
+            , return NoSnapshot
+            ]
+
+instance sc ~ TPraosMockCrypto
+      => Arbitrary (CardanoLedgerState sc) where
+  arbitrary = arbitraryHardForkState (Proxy @LedgerState)
+
+instance sc ~ TPraosMockCrypto
+      => Arbitrary (HardForkConsensusState (CardanoEras sc)) where
+  arbitrary = arbitraryHardForkState (Proxy @WrapConsensusState)
+
+-- | Forwarding
+instance Arbitrary (ConsensusState (BlockProtocol blk))
+      => Arbitrary (WrapConsensusState blk) where
+  arbitrary = WrapConsensusState <$> arbitrary
+
+-- | NOTE: Byron hashes are always 32 bytes, but for Shelley it depends on the
+-- crypto: with 'TPraosStandardCrypto' the hash is also 32 bytes, but with
+-- 'TPraosMockCrypto' it is 4 bytes ('ShortHash'). As this generator is only
+-- used in the roundtrip tests, where we don't care about the length of the
+-- hash, this difference is fine.
+--
+-- When we would store them on disk (specifically, in the secondary index of
+-- the ImmutableDB) and try to read them again, this would go wrong, as we
+-- require all hashes to have the same length.
+instance Crypto sc => Arbitrary (OneEraHash (CardanoEras sc)) where
+  arbitrary = OneEraHash <$> oneof
+    [ toRawHash (Proxy @ByronBlock) <$> arbitrary
+    , toRawHash (Proxy @(ShelleyBlock sc)) <$> arbitrary
+    ]
+
+instance Arbitrary (AnnTip (CardanoBlock TPraosMockCrypto)) where
+  arbitrary = oneof
+    [ mapAnnTip TipInfoByron   <$> arbitrary @(AnnTip (ByronBlock))
+    , mapAnnTip TipInfoShelley <$> arbitrary @(AnnTip (ShelleyBlock TPraosMockCrypto))
+    ]
+
+{-------------------------------------------------------------------------------
+  NodeToNode
+-------------------------------------------------------------------------------}
+
+instance TPraosCrypto sc => Arbitrary (HardForkNodeToNodeVersion (CardanoEras sc)) where
+  arbitrary =
+    elements $ NE.toList $ supportedNodeToNodeVersions (Proxy @(CardanoBlock sc))
+
+arbitraryNodeToNode
+  :: ( Arbitrary (WithVersion ByronNodeToNodeVersion byron)
+     , Arbitrary shelley
+     )
+  => (byron   -> cardano)
+  -> (shelley -> cardano)
```

Comment:
Parameterise over `f` instead.

Reply:
Hmm, all these coercions/proxies/unwrappings are going to be annoying and won't work for `SomeResult`, so I'm going to keep it this way (at least for now).